ext/mysql2/result.c

```diff
@@ -169,9 +178,161 @@ static VALUE rb_mysql_result_fetch_field(VALUE self, unsigned int idx, int symbo
+        } else if (field->flags & SET_FLAG) {
+          rb_field_type = rb_str_new_cstr("set");
+        } else {
+          if (field->charsetnr == MYSQL2_BINARY_CHARSET) {
+            rb_field_type = rb_sprintf("binary(%ld)", field->length);
+          } else {
+            rb_field_type = rb_sprintf("char(%ld)", field->length / MYSQL2_MAX_BYTES_PER_CHAR);
+          }
+        }
+        break;
+      case MYSQL_TYPE_VAR_STRING:   // char[]
+        if (field->charsetnr == MYSQL2_BINARY_CHARSET) {
+          rb_field_type = rb_sprintf("varbinary(%ld)", field->length);
+        } else {
+          rb_field_type = rb_sprintf("varchar(%ld)", field->length / MYSQL2_MAX_BYTES_PER_CHAR);
+        }
+        break;
+      case MYSQL_TYPE_VARCHAR:      // char[]
+        rb_field_type = rb_sprintf("varchar(%ld)", field->length / MYSQL2_MAX_BYTES_PER_CHAR);
+        break;
+      case MYSQL_TYPE_TINY_BLOB:    // char[]
+        rb_field_type = rb_str_new_cstr("tinyblob");
+        break;
+      case MYSQL_TYPE_BLOB:         // char[]
+        if (field->charsetnr == MYSQL2_BINARY_CHARSET) {
+          switch(field->length) {
+            case 255:
+              rb_field_type = rb_str_new_cstr("tinyblob");
+              break;
+            case 65535:
+              rb_field_type = rb_str_new_cstr("blob");
+              break;
+            case 16777215:
+              rb_field_type = rb_str_new_cstr("mediumblob");
+              break;
+            case 4294967295:
+              rb_field_type = rb_str_new_cstr("longblob");
+            default:
+              break;
+          }
+        } else {
+          if (field->length == (255 * MYSQL2_MAX_BYTES_PER_CHAR)) {
+            rb_field_type = rb_str_new_cstr("tinytext");
+          } else if (field->length == (65535 * MYSQL2_MAX_BYTES_PER_CHAR)) {
+            rb_field_type = rb_str_new_cstr("text");
+          } else if (field->length == (16777215 * MYSQL2_MAX_BYTES_PER_CHAR)) {
+            rb_field_type = rb_str_new_cstr("mediumtext");
+          } else if (field->length == 4294967295) {
+            rb_field_type = rb_str_new_cstr("longtext");
+          } else {
+            rb_field_type = rb_sprintf("text(%ld)", field->length);
+          }
+        }
+        break;
+      case MYSQL_TYPE_MEDIUM_BLOB:  // char[]
+        rb_field_type = rb_str_new_cstr("mediumblob");
+        break;
+      case MYSQL_TYPE_LONG_BLOB:    // char[]
+        rb_field_type = rb_str_new_cstr("longblob");
+        break;
+      case MYSQL_TYPE_BIT:          // char[]
+        rb_field_type = rb_sprintf("bit(%ld)", field->length);
+        break;
+      case MYSQL_TYPE_SET:          // char[]
+        rb_field_type = rb_str_new_cstr("set");
+        break;
+      case MYSQL_TYPE_ENUM:         // char[]
+        rb_field_type = rb_str_new_cstr("enum");
+        break;
+      case MYSQL_TYPE_GEOMETRY:     // char[]
+        rb_field_type = rb_str_new_cstr("geometry");
+        break;
+      default:
```

Comment:
If I understand it well it will return `nil` for unknown type, right?

Reply:
I wonder how we can force a unit test down the no-match path?

Reply:
If we know a unknown to test for it, would it be unknown anymore :laughing:?

Reply:
Anyway I've made it return 'unknown' for this case.

Reply:
I was looking into this @sodabrew and I have found out there's no easy way how to introduce custom column type in MySQL :/

But maybe there's some way how to create table including all MySQL data types (by some meta programming) and it would be possible to check if we cover all cases there.

Reply:
It will fail if we include `MYSQL_TYPE_JSON` though. I cannot find any code to handle `MYSQL_TYPE_JSON`.

Reply:
See #693 for the history on how JSON is supported; basically it's a fall-through to `char[]`.

It seems to me that the type table should be updated to explicitly include JSON so that it's not "accidentally" supported but is explictly routed through the `char[]` path of the result handler.

What is the failure you're seeing if `MYSQL_TYPE_JSON` is included in this table?